File: 2021/18.py
a = []


def parse_num(num: str):
    a = []
    for i in num:
        if i == '[':
            a.append(i)
        elif i == ']':
            a.append(i)
        elif i == ',':
            continue
        else:
            a.append(int(i))
    return a


def explode(a):
    d = 0
    for i in range(len(a)):
        if a[i] == '[':
            d += 1
        elif a[i] == ']':
            d -= 1
        elif isinstance(a[i], int):
            if d == 5:
                first = a[i]
                second = a[i+1]
                b = a[:i-1]
                j = i-2
                while not isinstance(b[j], int):
                    j -= 1
                    if j < 0:
                        break
                if j > 0:
                    b[j] += first
                b.append(0)
                b.extend(a[i+3:])
                j = i
                while not isinstance(b[j], int):
                    j += 1
                    if j >= len(b):
                        break
                if j < len(b):
                    b[j] += second
                return True, b
    return False, a


def split(a):
    for i in range(len(a)):
        b = a[i]
        if isinstance(b, int):
            if b >= 10:
                c = a[:i]
                c.extend(['[', b//2, (b+1)//2, ']'])
                c.extend(a[i+1:])
                return True, c
    return False, a


def reduce(a):
    reduced = True
    while reduced:
        reduced, a = explode(a)
        if not reduced:
            reduced, a = split(a)
    return a


def get_magnitude(a):
    i = 0
    while len(a) > 1:
        if isinstance(a[i], int) and isinstance(a[i+1], int):
            mag = 3 * a[i] + 2 * a[i+1]
            del a[i-1:i+3]
            a.insert(i-1, mag)
            i -= 2
        else:
            i += 1
    return a[0]


def add(a, b):
    c = ['[']
    c.extend(a)
    c.extend(b)
    c.extend(']')
    return c


a = input()
a = parse_num(a)
numbers = [a]
while True:
    b = input()
    if b == "exit":
        break
    b = parse_num(b)
    numbers.append(b)
mag = 0
for i in range(len(numbers)):
    for j in range(len(numbers)):
        if i == j:
            continue
        x = reduce(add(numbers[i], numbers[j]))
        mag = max(mag, get_magnitude(x))
# print(a)
# print(get_magnitude(a))
print(mag)

File: 2021/18.cpp
#include <bits/stdc++.h>

using namespace std;

string add(string a, string b)
{
    return '[' + a + ',' + b + ']';
}

bool explode(string &a)
{
    int d = 0;
    for (int i = 0; i < a.size(); i++)
    {
        if (a[i] == '[')
            d++;
        if (a[i] == ']')
            d--;
        if (!isdigit(a[i]))
            continue;
        pair<int, int> p = make_pair(-1, -1);
        int k = i;
        string s = "";
        while (isdigit(a[k]))
        {
            s += a[k];
            k++;
        }
        if (s != "")
            p.first = stoi(s);
        s = "";
        k++;
        while (isdigit(a[k]))
        {
            s += a[k];
            k++;
        }
        if (s != "")
            p.second = stoi(s);
        if (d == 5 && p.first != -1 && p.first != -1)
        {

            a = a.substr(0, i - 1) + '0' + a.substr(i + 4);
            for (int j = i - 2; j >= 0; j--)
            {
                if (isdigit(a[j]))
                {
                    int k = j;
                    string s = "";
                    while (isdigit(a[k]))
                    {
                        s += a[k];
                        k++;
                    }
                    a = a.substr(0, j) + to_string(p.first + stoi(s)) + a.substr(j + 1);
                    break;
                }
            }
            for (int j = i; j < a.size(); j++)
            {
                if (isdigit(a[j]))
                {
                    int k = j;
                    string s = "";
                    while (isdigit(a[k]))
                    {
                        s += a[k];
                        k++;
                    }
                    a = a.substr(0, j) + to_string(p.second + stoi(s)) + a.substr(j + 1, INT_MAX);
                    break;
                }
            }
            return true;
        }
    }
    return false;
}

bool split(string &a)
{
    for (int i = 0; i < a.size() - 1; i++)
    {
        if (isdigit(a[i]) && isdigit(a[i + 1]))
        {
            int c = 10 * (a[i] - '0') + a[i + 1] - '0';
            a = a.substr(0, i) + '[' + to_string(c / 2) + ',' + to_string((c + 1) / 2) + ']' + a.substr(i + 2, INT_MAX);
            return true;
        }
    }
    return false;
}

void reduce(string &a)
{
    while (true)
    {
        bool exploded = true;
        while (exploded)
            exploded = explode(a);
        bool splitted = split(a);
        if (!splitted)
            break;
    }
}

int main(int argc, char const *argv[])
{
    string a;
    cin >> a;
    reduce(a);
    while (true)
    {
        string b;
        cin >> b;
        if (b == "exit")
            break;
        reduce(b);
        a = add(a, b);
        reduce(a);
    }
    cout << a << endl;
    return 0;
}
